### ISSUE TITLE
Make Linux and OSX binaries executable on export

### ISSUE THREAD

**Issue by voidplayer** [godotengine/godot#527]
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

file doesnt get proper +x permission on export

**Comment by akien-mga** [comment]
I can confirm this is still the case on the current  bd736e5.

**Comment by Cheeseness** [comment]
This happens even if the template binary has +x

Adding an export_project() function to platform/x11/export/export.cpp that contained a OS::execute() call to run chmod would probably be the way to go, but I'm not familiar enough with Godot's codebase yet to help here.

**Comment by akien-mga** [comment]
Adding the ``chmod +x`` call itself is relatively trivial:
```
diff --git a/tools/editor/editor_import_export.cpp b/tools/editor/editor_import_export.cpp
index bdf1726..712bc72 100644
--- a/tools/editor/editor_import_export.cpp
+++ b/tools/editor/editor_import_export.cpp
@@ -1372,6 +1372,11 @@ Error EditorExportPlatformPC::export_project(const String& p_path, bool p_debug,
                }
        }
 
+       List<String> args;
+       args.push_back("+x");
+       args.push_back(p_path);
+       OS::get_singleton()->execute("/usr/bin/chmod",args,false);
+
        String dstfile = p_path.replace_first("res://","").replace("\\","/");
        if (export_mode!=EXPORT_EXE) {
```

There are several problems though:
- the ``EditorExportPlatformPC`` exporter is used for both Windows and Linux templates, and we don't want to ``chmod +x`` the Windows binaries
- obviously, we can't use ``chmod`` when exporting Linux binaries on a Windows host
- all in all, the gain is little because the executable permission is set for the current system, so if you send the binary to another Linux user, they would receive it with permission 644. The workaround is to put the binary in an archive, which preserves the execution bit.

All in all, I'd be for just making a nice docs page explaining that Linux binaries have to be made executable for distribution.

**Comment by nunodonato** [comment]
unrelated but related, does anyone know if the exported mac binary also has the same problem? I guess the file in inside the .app package needs to have the executable bit set too, right? (its still unix)

**Comment by Cheeseness** [comment]
> the executable permission is set for the current system

Just in case this hasn't been communicated clearly, Godot on Linux doesn't preserve +x on Linux binaries during export even if the template binaries have +x.

**Comment by akien-mga** [comment]
> unrelated but related, does anyone know if the exported mac binary also has the same problem? 

The binary is made executable before it's packaged in the .app archive (which is just a .zip, and should preserve permissions). There was an unrelated issues though that @punto- recently fixed: 6962fd91

> Just in case this hasn't been communicated clearly, Godot on Linux doesn't preserve +x on Linux binaries during export even if the template binaries have +x.

Because, as I said, the execution permission is registered in your own system and not in the file itself. As the binary is read and written again in another location, these permissions are lost and would have to be manually reapplied.

I'm not sure what's the reason for rewriting the full binary instead of just copying it though.

**Comment by Cheeseness** [comment]
> The binary is made executable before it's packaged in the .app archive (which is just a .zip, and should preserve permissions).

For clarity, .app packages are a folder structure rather than an archive. Mac users still need to extract the .zip before they can launch the .app. You're right that it's in a .zip to preserve the permissions (although as far as I'm aware, that's a result of various unzip implementations respecting [this stuff](http://unix.stackexchange.com/questions/14705/the-zip-formats-external-file-attribute) rather than something that's identifiably a part of the zip format).

The same strategy could be viable for making it easier for Windows based developers to ship Godot games on Linux.

> Because, as I said, the execution permission is registered in your own system and not in the file itself. As the binary is read and written again in another location, these permissions are lost and would have to be manually reapplied.

Correct. Your comments about sending it to another Linux user seemed to be implying that the exported binary on the same system (computer) would retain +x permissions, but I see now that you were meaning that it's stored at the file system level. Apologies for misunderstanding.

> I'm not sure what's the reason for rewriting the full binary instead of just copying it though.

I'm not certain either. Could it be that the export code for making a single binary (containing the game's data) was written first and reused for the pack modes that don't modify the binary?

**Comment by akien-mga** [comment]
Looks like on Linux we could use `chmod` directly from `<sys/stat.h>` directly: http://stackoverflow.com/a/12694263
I guess OSX might have chmod too, though this needs checking.

For Windows, we'll need to find another API, but I'm pretty sure something exists. Any idea about it @vnen?

**Comment by punto-** [comment]
For osx, the binaries are read from the template and used to create a .zip file aren't they? We don't actually generate the final .app structure on the filesystem and then zip it, we create the .zip programatically. The solution there is to turn on the executable bit for the file inside the zip. That should be portable, we could do it from any OS (if such a bit even exists)

**Comment by akien-mga** [comment]
> The solution there is to turn on the executable bit for the file inside the zip. That should be portable, we could do it from any OS (if such a bit even exists)

That's the idea, but we need to find the APIs that allow to do this for all platforms/executable types.

**Comment by punto-** [comment]
But the point is this an api in the zip library, it shouldn't be platform
specific.. It's up to the unzip program in the target OS to actually make
it executable when it gets unzipped, if the filesystem supports it.

**Comment by akien-mga** [comment]
Well, that's worth looking into for OSX if we can set the executable bit when zipping, but for Linux binaries it won't work as we don't zip them on export.

**Comment by Cheeseness** [comment]
That's what Godot is already doing for OSX. I think @punto- is suggesting that Linux exports be deployed to zip via a similar method as well.

**Comment by akien-mga** [comment]
For OSX/zip, we can maybe look at the file attributes in https://github.com/godotengine/godot/blob/master/core/io/zip.h#L100-L108

See e.g. http://unix.stackexchange.com/questions/14705/the-zip-formats-external-file-attribute

That still wouldn't solve the issue of setting the +x when copying the Linux templates (we seem to read the Linux binaries and rewrite it to the chosen destination, thus losing the existing executable bit), but it would be great already to have working OSX .app templates.

**Comment by akien-mga** [comment]
@Cheeseness Well it doesn't seem to be done so far for OSX, see https://github.com/godotengine/godot/issues/4618#issuecomment-218457090
I don't think that we should start zipping Linux binaries just for that either.

**Comment by Cheeseness** [comment]
https://github.com/godotengine/godot/issues/4618#issuecomment-218457090 isn't in line with what I recall experiencing when working on the two Godot projects I've shipped for Mac OS. I'll double check that next week when I have some more time.

IMO, offering a zipped Linux binary export would be valuable for Windows based developers who want to ship Linux builds.

**Comment by akien-mga** [comment]
Some more reading about external file attributes in the ZIP API (it's crazy how undocumented this stuff can be o_O): http://stackoverflow.com/a/6297838

My understanding is that we could add something like that after: https://github.com/godotengine/godot/blob/master/platform/osx/export/export.cpp#L401
```
if (file == pkg_name+".app/Contents/MacOS/"+binary_to_use) {
    fi.external_fa |= 0755 << 16L # -rwxr-xr-x
}
```

Which would set the UNIX permissions that we want on the binary without (a priori) messing with the DOS permissions.

**Comment by vnen** [comment]
For what I can gather, the ZIP approach is the only way to provide Linux binaries on Windows with the proper execute bit. If one extract this file on Windows, the NTFS will likely trash Unix permission bits. IME moving files from Windows to Linux always made them 777.

**Comment by Cheeseness** [comment]
> IME moving files from Windows to Linux always made them 777.IME moving files from Windows to Linux always made them 777.

What method are you using to transfer them? This has never ever happened for me.

The only thing that has that kind of behaviour that I know of is Steam, and that's the client setting 777 on all files it downloads (if you listen quietly, you can hear the sound of a million security conscious people facepalming every time a download finishes) rather than anything inherent to the file transfer protocol.

**Comment by akien-mga** [comment]
I'll give a try at the OSX part this week-end using what I've documented above. If it works, I'll see what can be done for Linux.

**Comment by vnen** [comment]
Actually, my problem with permissions was not from *moving* files between OSs, but with files *created* on Windows and then transferred to Linux. And that's what the exporter does, so...

**Comment by Cheeseness** [comment]
@vnen How are you transferring them though? If you're mounting a Windows drive on Linux and it gets mounted with 777 permissions, then that's sidestepping the issue in a way that's not so relevant to the issue at hand.

@akien-mga As mentioned in IRC, I've needed to give +x to bins on Mac OS for at least one of my previous projects, so I'd been remembering wrong. Apologies!

**Comment by akien-mga** [comment]
I've had a go at adding the fix in the zipping part for OSX, but so far it didn't work :(

I tried:
```diff
diff --git a/platform/osx/export/export.cpp b/platform/osx/export/export.cpp
index 47b0392..cdfb657 100644
--- a/platform/osx/export/export.cpp
+++ b/platform/osx/export/export.cpp
@@ -399,6 +399,9 @@ Error EditorExportPlatformOSX::export_project(const String& p_path, bool p_debug
                        fi.dosDate=info.dosDate;
                        fi.internal_fa=info.internal_fa;
                        fi.external_fa=info.external_fa;
+                       // Add UNIX exec permission to binary
+                       if (file == pkg_name+".app/Contents/MacOS/"+pkg_name)
+                               fi.external_fa |= 0755 << 16; // -rwxr-xr-x
 
                        int err = zipOpenNewFileInZip(dpkg,
                                file.utf8().get_data(),
```

But it had no visible effect. (I made sure that the if block is valid ;)).

Then I tried something based on how it was implemented in QuaZip: https://sourceforge.net/p/quazip/code/155/tree//trunk/quazip/quazip/quazipnewinfo.cpp?diff=50cc5aa72718464452efcd66:154
```
diff --git a/platform/osx/export/export.cpp b/platform/osx/export/export.cpp
index 47b0392..cdfb657 100644
--- a/platform/osx/export/export.cpp
+++ b/platform/osx/export/export.cpp
@@ -399,6 +399,9 @@ Error EditorExportPlatformOSX::export_project(const String& p_path, bool p_debug
                        fi.dosDate=info.dosDate;
                        fi.internal_fa=info.internal_fa;
                        fi.external_fa=info.external_fa;
+                       // Add UNIX exec permission to binary
+                       if (file == pkg_name+".app/Contents/MacOS/"+pkg_name)
+                               fi.external_fa = (fi.external_fa & ~0xFFFF0000u) | (0100755 << 16); // -rwxr-xr-x
 
                        int err = zipOpenNewFileInZip(dpkg,
                                file.utf8().get_data(),
```

But this seemed to corrupt the filename:
```
$ unzip Tatata21.zip 
Archive:  Tatata21.zip
  inflating: Tatata21.app/Contents/Resources/icon.icns  
   skipping: Tatata21app/Contents/MacOS/Tatata21  volume label   // "volume label"?
  inflating: Tatata21.app/Contents/PkgInfo  
  inflating: Tatata21.app/Contents/Info.plist  
  inflating: Tatata21.app/Contents/Resources/data.pck  
```

I guess we're on the right track, but someone who understand a thing about manipulating bits in C++ might find the proper syntax :)

**Comment by chenzhekl** [comment]
@akien-mga Hi, I would like to help. But it seems everything in `platform/osx/export` is commented out. Could you give some hints where I can start the work?

**Comment by volzhs** [comment]
@chenzhekl you can take a look at [2.1 branch](https://github.com/godotengine/godot/tree/2.1)

**Comment by marcelofg55** [comment]
I've been digging about the zip issue when trying to add the 0755 attributes.
It seems that our minizip thirdparty code is the problem.

Image of the preview of a zp file created by our export code (note the lack of attributes in the last column):
![screen shot 2017-09-17 at 4 06 25 pm](https://user-images.githubusercontent.com/10578225/30524101-982bb746-9bc3-11e7-8509-fbcfe25e2032.png)

Image of a zip file created by an external zipper (the the correct attributes in the last column):
![screen shot 2017-09-17 at 4 06 17 pm](https://user-images.githubusercontent.com/10578225/30524114-d2a4d81c-9bc3-11e7-9159-91d105a4792a.png)

**Comment by BastiaanOlij** [comment]
For building the Godot 3 Mac exporter I did add the chmod to make the executable, uhm, executable. In Godot 3 provided you are building on a Mac, it'll build both the application bundle, optionally code sign and build a DMG out of the app bundle. If you export on linux it still zips it up and can't code sign so it'll probably suffer from the issue described above.

Anyway, have a look at how things work now in Godot 3, might be possible to retrofit to Godot 2.x

**Comment by akien-mga** [comment]
I think @BastiaanOlij's existing code should just be rid of the `#ifdef OSX_ENABLED` that restricts it to exporting from OSX, and use @marcelofg55's new `OS::_chmod` instead.

**Comment by BastiaanOlij** [comment]
I'm just going to tag @endragor in here, he has taken my solution much further in recent months on Godot 3.

**Comment by akien-mga** [comment]
Moving to 3.1 again for a full resolution... it's much better than in 2.1 already but still inconsistent between platforms.

**Comment by exts** [comment]
Still an issue when exporting linux projects, permissions don't stick when distributing the build. Maybe a solution could be having an option to `tar.gz` a linux build for redistribution to keep execution permissions. I've gotten reports where users have to chmod +x every time, even when I test builds on my other desktop (ubuntu 18)

**Comment by aaronfranke** [comment]
What else needs to be done after https://github.com/godotengine/godot/pull/11362 ? What's missing?

**Comment by akien-mga** [comment]
> What else needs to be done after #11362 ? What's missing?

The feature should work when exporting Linux and macOS binaries from Linux or macOS, but this doesn't cover the export from Windows, which has no notion of Unix permissions. For Windows, the solution would be to zip everything and set the Unix permissions in the zip, but I failed to do it last time I attempted it.

**Comment by marcelofg55** [comment]
> The feature should work when exporting Linux and macOS binaries from Linux or macOS, but this doesn't cover the export from Windows, which has no notion of Unix permissions. For Windows, the solution would be to zip everything and set the Unix permissions in the zip, but I failed to do it last time I attempted it.

It doesn't work on any platform last time I checked, minizip is the problem, check this reply: https://github.com/godotengine/godot/issues/527#issuecomment-330075877

**Comment by duckinator** [comment]
I have nightly builds for my game, exported from Godot 3.1.1 on Linux, for Windows + Linux + macOS and haven't ran into any issues, so this appears to be resolved at least when exporting on Linux.

**Comment by Anutrix** [comment]
Windows Subsystem for Linux (WSL) on Windows 10 can be used to solve this but it's optional feature and not everyone installs it.

> 
> 
> > The feature should work when exporting Linux and macOS binaries from Linux or macOS, but this doesn't cover the export from Windows, which has no notion of Unix permissions. For Windows, the solution would be to zip everything and set the Unix permissions in the zip, but I failed to do it last time I attempted it.
> 
> It doesn't work on any platform last time I checked, minizip is the problem, check this reply: [#527 (comment)](https://github.com/godotengine/godot/issues/527#issuecomment-330075877)

**Comment by aaronfranke** [comment]
@marcelofg55 Have you opened an upstream bug report with the minizip developers?

**Comment by bruvzg** [comment]
>  Have you opened an upstream bug report with the minizip developers?

`minizip` can set permissions, but unless "made by" field is populated with the specific version with Unix permission support, "unzip"s ignore all permissions. See [#33447 > export.cpp > L729-746](https://github.com/godotengine/godot/pull/33447/files#diff-d7457c00902d2bf7783035de6db0f164R729-R746).

At least on macOS, executable extracted form the ZIP created in this way have correct permissions set. I have tested creating ZIPs on macOS and Windows. And extracting it on macOS using Archiver (system default extractor), Keka and Midnight Commander.

**Comment by akien-mga** [comment]
@bruvzg Nice catch! So #33447 solves the OSX part of this issue I think, we'd just need to add ZIP export for Linux with similar logic I suppose?

**Comment by bryantfhayes** [comment]
Is this issue resolved in 3.2.1? Or is it targeting 4.0?

**Comment by Calinou** [comment]
@bryantfhayes It's targeting 4.0 at least.

**Comment by pouleyKetchoupp** [comment]
I've been investigating this issue when exporting as zip from Windows for OSX and got some intriguing results. Depending on which utility is used on Mac to uncompress the zip file, it works in some cases.

On my custom branch based on 3.1:
Using `unzip`: binary not set as an executable
Using `Archive Utility`: permissions are set correctly and the game launches

After merging #33447:
It's the other way around.
Using `unzip`: permissions are set correctly and the game launches
Using `Archive Utility`: binary not set as an executable

I'm going to try and investigate the differences in how flags are set when zipping files before and after this PR to see if there's a way to make it work in all cases, but in the meantime if someone has an idea about what could be happening I would be interested in any hint.

**Comment by TerraDev64** [comment]
I have the same problem.

**Comment by Hamzailer** [comment]
I also have this problem

**Comment by EamonnMR** [comment]
Is there a workaround to fix the app on a mac after we unzip it?

**Comment by akien-mga** [comment]
Use 3.2.3-stable.

**Comment by sszigeti** [comment]
Just exported a game for Linux on Windows10 with 3.3.stable, and it didn't add the +x permission to the executable.

Also, perhaps unrelated, but the exported Mac version (exported from a Mac with 3.3.stable) doesn't start using the **downloaded** .dmg, but the same, local .dmg starts without issues.

**Comment by sairam4123** [comment]
>>> Also, perhaps unrelated, but the exported Mac version (exported from a Mac with 3.3.stable) doesn't start using the downloaded .dmg, but the same, local .dmg starts without issues.

Any problems with that, my friend also has the similar problem do I have to use 3.2.3.stable?

**Comment by bruvzg** [comment]
> Any problems with that, my friend also has the similar problem do I have to use 3.2.3.stable?

Is it signed? Unsigned macOS apps won't run when downloaded from unknown source, until quarantine file attribute is removed (use `xattr -dr com.apple.quarantine /path/to/Application.app` command after downloading). On Apple Silicon macs unsigned apps won't run at all.

**Comment by sairam4123** [comment]
No it's not signed, removing the com.apple.quarantine worked!

**Comment by bruvzg** [comment]
Currently, the situation for macOS exports looks like this:

- App is signed with Apple issues signature and notarized - it will display warning dialog on the first start and run without issues.
- App is ad-hoc signed (can be done by selecting `Codesign/Enabled` in the export options, and setting `Codesign/Identity` to `-`) - app won't run, but you can unlock it with right-clicking app in the Finder and selecting "Open" two times in the row.
- App is not signed - It won't run and can be unlocked using the aforementioned command, for older Godot versions without Apple Silicon support right-clicking method might also work.

See https://disable-gatekeeper.github.io/ for some extra options.

**Comment by Anutrix** [comment]
Some of bug comments here might have been fixed with https://github.com/godotengine/godot/pull/62616 in master.

Also, wasn't this ticket initially about `+x` attribute but now is blocked only by gatekeeper stuff? Should be close and make a new ticket for this?

Also, Gatekeeper situation can only solved completely by Apple notarization.
Possible actions that can be taken/discussed for Gatekeeper:
 - Some doc notes mentioning is Gatekeeper related sources/solutions can be added(including bruvg's [comment](https://github.com/godotengine/godot/issues/527#issuecomment-872184727)).
 - If possible, do it apparently how [Blender](https://developer.blender.org/rBe60a01483ff048ef07fe7c148a5b8f33ed86d15d#change-LScX2UOSaP1M) does it.

**Comment by aaronfranke** [comment]
@Anutrix No, I can confirm that on macOS 12.4 with a recent Godot master build that it's not setting executable permission on export. When I export the `.app` and try to run, macOS says "The application “myapp” can’t be opened". If I run `chmod -R +x mygame.app` then it works fine. This issue is still present, Godot isn't ensuring the app is executable.

**Comment by akien-mga** [comment]
That was a regression in alpha 11, that's fixed already.

**Comment by voylin** [comment]
@Calinou Should this not have the label of platform:macos instead of Linuxbsd?
On Linux with Godot 4.1.1, export are executable on export for me.

**Comment by akien-mga** [comment]
Discussed this briefly with @bruvzg, this should be resolved as far as technically feasible by now (probably since at least 4.0, but there might have been further improvements in 4.1 so I'll mark it as fixed in 4.1).

The executable permissions for Linux and macOS binaries are set in all cases where it's possible.

Notably on Windows, which doesn't have a concept of file permissions, the permissions can only be set when doing an export as `.zip`, which exports a zip with the game binary and PCK together, and sets the permissions in the zip. These permissions should work as long as you don't unzip it and rezip it manually on Windows.

**Comment by Calinou** [comment]
> Note that the UX to do a zip export isn't great currently, for example for a Linux build you should do a normal export as usual, and replace the gamename.x86_64 by gamename.zip (you can see that the filter mentions both .x86_64 and .zip as valid extensions).

To clarify, this is **not** the same as using the **Export PCK/ZIP** option. You need to use the standard **Export Project** button instead.
